### PR TITLE
[VectorDistribution] Add distribution pattern for vector::MultiDimReductionOp

### PR DESCRIPTION
**Pull request by Groverkss**
This patch adds support for distribution vector::MultiDimReductionOp. The distribution works in two steps:

- Local reduction: add a vector.multi_reduction to reduce across batches, outers, element
- Subgroup reduction: Use gpu.shuffle to reduce among threads

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp]: nit: I think you can use `std::exclusive_scan` here. If not, the pattern for reverse iteration with unsigned index is: `for (size_t e = x.size(), i = e - 1; i < e; --i)`

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp]: ubernit
```suggestion
    auto accVector = dyn_cast<VectorValue>(multiReduceOp.getAcc());
```

also below

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp]: Can you add this in this PR? I have a feeling this may bite us if we ever try to enable vector distribution on a new target.

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp]: Can you file an issue against llvm-project? I will add it to the gpu dialect when I have some spare cycles.

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp:617]: Here we assume that all lanes are active? Is this always the case or can it happen that we don't evenly distribute across subgroups?

File: compiler/src/iree/compiler/Codegen/Common/GPU/test/gpu_nested_layout_vector_distribution.mlir
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Common/GPU/test/gpu_nested_layout_vector_distribution.mlir:1025]: Can you add a comment explaining what is the shuffle width and how many elements we reduce per thread?

File: compiler/src/iree/compiler/Codegen/Common/GPU/test/gpu_nested_layout_vector_distribution.mlir
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Common/GPU/test/gpu_nested_layout_vector_distribution.mlir]: nit: Should we align the one above the same way?

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp]: Instead of applying the full permutation
```suggestion
  return activeThreadStrides[layout.getThreadOrder()[dim]];
```

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp]: `MultiDimReduction` returns a scalar if and only if not all of the dimensions are reduced, and that is already guaranteed with the earlier check so this can just be an assert. Best practice is to avoid pattern failures after generating IR also.

https://mlir.llvm.org/docs/Dialects/Vector/#vectormulti_reduction-vectormultidimreductionop

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp]: Best practice is to not generate any IR before failing a pattern. Can you move this check to the beginning of the whole pattern?

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp:491]: It would be nice to have a short doc here describing the lowering steps (local reduce along batch/outer/element dims followed by a shuffle partial reduce across the distributed dims).

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp
Comment (qedawkins) [reply compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp]: +1, I've done the same before and it ends up being more hassle than it's worth, so doing the extra plumbing up front is much appreciated. FWIW we'll need to change this pretty soon for gfx1100 where we are using subgroup size 32 right now.

File: llvm-external-projects/iree-dialects/lib/Dialect/VectorExt/IR/VectorExtAttrs.cpp
Comment (qedawkins) [inline llvm-external-projects/iree-dialects/lib/Dialect/VectorExt/IR/VectorExtAttrs.cpp:322]: @raikonenfnu this might have been the issue we discussed the other day fyi.

File: compiler/src/iree/compiler/Codegen/Common/GPU/test/gpu_nested_layout_vector_distribution.mlir
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/Common/GPU/test/gpu_nested_layout_vector_distribution.mlir:1056]: nit: Can you make this first test show the actual resulting distributed vector shapes? The lowering is fairly involved and it helps to have at least one test that is a bit verbose.

**Review by qedawkins — APPROVED** [review]
LGTM after addressing comments. Wait for approval from Jakub also.

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp
Comment (Groverkss) [reply compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp]: Is the issue you created in IREE enough? https://github.com/iree-org/iree/issues/17082 Maybe we can move this to llvm instead.

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp
Comment (Groverkss) [reply compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp:617]: We don't assume this. getShuffleOffset creates the strides in a way that inactive lanes are accounted for.

File: compiler/src/iree/compiler/Codegen/Common/GPU/test/gpu_nested_layout_vector_distribution.mlir
Comment (Groverkss) [reply compiler/src/iree/compiler/Codegen/Common/GPU/test/gpu_nested_layout_vector_distribution.mlir:1025]: added

File: compiler/src/iree/compiler/Codegen/Common/GPU/test/gpu_nested_layout_vector_distribution.mlir
Comment (Groverkss) [reply compiler/src/iree/compiler/Codegen/Common/GPU/test/gpu_nested_layout_vector_distribution.mlir]: done

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp
Comment (Groverkss) [reply compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp:491]: done

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp
Comment (Groverkss) [reply compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp]: I plumbed the subgroup size to ContractionVectorLayoutOptions. Not sure how to plumb it further.

Comment (qedawkins) [reply compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp]: Nice, all you should need now is just retrieving the subgroup size from the enclosing function. You can get the subgroup size with this helper: https://github.com/iree-org/iree/blob/afd7cab58be44c035eb80aaad39652b4f5d5a667/compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenAttrs.h#L51

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp
Comment (kuhar) [reply compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp]: Yeah I thought about a more specific issue filed against llvm-project

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp
Comment (Groverkss) [reply compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp]: Done for vector distribution pass. For transform dialect stuff though i hardcoded the subgroup size for now.

File: tests/e2e/matmul/generate_e2e_matmul_tests.py
Comment (kuhar) [inline tests/e2e/matmul/generate_e2e_matmul_tests.py]:
```suggestion
                # TODO: This is only valid for gfx9. Change this for RDNA3
```

**Review by kuhar — APPROVED** [review]
LGTM!

**Pull request merged**
